### DISCUSSION TITLE
MQTT Broadcast

### DISCUSSION THREAD

**Discussion by YPOC** [NightscoutFoundation/xDrip#2528] [Ideas]
Hello,

is it possible to broadcast glucose values to an MQTT Broker?

MQTT is a lightweight protocol meant for IOT communication, where multiple clients can broadcast to a broker and can simultaneously subscribe to new messages published to this broker. Implementation would be perfect for my target of having my glucose displayed on other devices such as desktop pc while reducing the dependency on Nightscout.

**Comment by Navid200** [comment]
May I ask why dependency on Nightscout is an issue?

We already have put a lot of Nightscout support into xDrip.

**Reply by YPOC** [reply]
For me personally, I don't require Nightscout most of the time. There are only two exceptions:
I use it to generate reports for my doctor's appointment every 3-6 months with zreptil's reporter tool. But here I don't interact with Nightscout directly. I wish it was possible to access AAPS's database directly to generate those reports.
Secondly I use Delubear's GlucoseTray https://github.com/Delubear/GlucoseTray to show my current glucose reading in Windows' task bar. If xDrip was able to publish to MQTT there'd be no more need for Nightscout for me.

With Heroku's drop of the free plan it has now become much more difficult to keep a free instance of Nightscout running, as all services require a credit card for identification, which I sadly don't have. Also it seems creating a Nightscout instance is still a hurdle for many.
I'm not asking to abandon Nightscout since I'm sure there are many uses for it - but simply not for me. I'm only asking for an alternative communication protocol.

**Reply by jonprogers** [reply]
I've been using MQTT for years by having a script that queries my mongo database every 5 mins and posts the bits I want as MQTT messages. Unfortunately, for your use case, this relies on nightscout as the method to get data from xDrip to mongo db. I appreciate that the ns API might be a better way of doing this now that I understand that too but I haven't wanted to rewrite a working script!!

xDrip does what it does reliably and really well on a device that may have patchy connection and manages ns upload and backfill through connection issues, my script benefits from all that clever mature code without me having to do anything, for which I am very grateful.

Personally I like nightscout and mongo being there, despite all my day to day treatment decisions being made by consulting xDrip, as I wouldn't like all my history to be held on a mobile device that I can break, lose or have stolen - although you may have other ways of protecting against this... 

My preference, from a data retention perspective, is for the mobile device (xdrip) to upload to one thing that is always available on the internet somewhere (such as ns and mongo) and then have other things know that they can connect to that reliably. Things are so much more reliable these days though that my concerns may be totally unfounded...

Interesting discussion you've started @YPOC , I'll be really interested in reading other replies.

**Reply by Navid200** [reply]
The requirement for a credit card existed even for Heroku even when it was free.
Nevertheless, I can see a legitimate case for someone who has no credit card.
Thanks for bringing this up.  I have asked and would like to consider options.

Please let me explain my concerns.
xDrip has become extremely complicated.
There is one individual who knows almost everything in xDrip.  He is always busy fixing things.  Every new feature will bring with it the requirements of fixing bugs.  
As a result, I suspect he gets thousands of emails a day.  This is not a reasonable workload for a volunteer.  It's not sustainable.

I have taken the role of the "bad guy" here (myself without anyone asking me to) trying to explain to everyone that we cannot keep adding features to xDrip.  
At the end, a reasonable approach is to minimize the number of new features to those that are absolutely necessary.

Again, thanks for bringing this up and we will be talking about it.

**Reply by Stan-Di** [reply]
I join the request. Please add the ability to send glucose data to mqtt server. It's very convenient and don't need to use Nightscout

**Reply by Navid200** [reply]
xDrip already offers many protocols for sending data.
It let's you use Nightscout, xDrip sync, Tidepool and others.

It may help if you explain what you need to do.  For example, if the intent is to have someone follow you, you can use xDrip sync.
But then, I would be curious what happens if you lose your phone?  Where will you get all your data back from?

If we spend the time and add a new protocol and only one or two people use it, it wouldn't be a very good outcome.

To help avoid this, please explain what you need to do that none of the existing protocols can do.
For example, why can you not use Nightscout?
I admit that setting up Nightscout requires spending some time.  But, you only need to do that once.

**Reply by Stan-Di** [reply]
Hello! Thanks for the answer! Mqtt is the best and simplest protocol for transmitting sensor readings (from temperature and radiation to glucose). It does not depend on Google servers like xDrip sync (for example, one of the phones did not have google service and I could not launch them, I got a brick), does not depend on the “business model” of Tidepool and other IFTTTs that recently changed the conditions again, does not depends on Nightscout, that hosters also constantly change conditions.

MQTT servers, of course, also can change conditions, but the same Mosquitto server is much easier to launch and connect with Home Assistant, for example, or any MQTT client, or Grafana etc. Using MQTT I can output sugar data to DIY watch with ESP8266 and OLED display.

Of course, I can come up with a crutch: run Nigthscout with MongoDB on the phone via Termux (this has already been done) and come up with a script to read data from Nightscout and send it to MQTT. But this is already a perversion. And the MQTT client library for any platform is open and free.

And after talking with people, I also realized that Nightscout’s functionality is redundant for many.

**Reply by Navid200** [reply]
I am going through old outstanding issues and am trying to fix them.
There are developers who add something to xDrip and then leave and never show up again.  The bugs that come up later are still there.
It is very unfair to consider this to be a reasonable approach.
It takes me much longer to figure out what's wrong and how to fix sompared to the person who added the feature.

But, since we are all volunteers, I cannot demand someone who added WearOS to xDrip to come and fix it now as an example.
So, either I have to fix it or it will remain there as a well-known shortcoming of xDrip.  

There is always going to be something that can be added.  But, we cannot just keep adding.
Nightscout already has related bugs we have to fix.  

An Android phone without Google play service?  What percentage of Android phones don't have that?
You can run Nightscout on a Google Cloud virtual machine.  Then, your database will be backed up on Nightsocut as well.
If you lose your phone, you will not lose all your data history.

**Reply by Stan-Di** [reply]
I'm not asking you to make a quick decision, I'm just asking you to consider the possibility. At the same time, I contacted the developer of a similar application and it turned out that it was easier to implement it there, via TCP connect.

And about Google, etc. Unfortunately, recent years have shown the politicization of some of the company’s actions: for example, sanctions for Huawei (which is why many phones do not have Google service). In Russia, for example, Google Cloud, Oracle and MongoDB Atlas no longer work. That's why I ran 100+ copies of Nightscout on my personal computer and am giving them away for free, but that's a whole different story. I would like diabetes to be out of politics, but unfortunately not everything depends on us...

Thank you for your answers, I will be glad if this opportunity appears someday!

**Reply by Navid200** [reply]
I don't want diabetes control to have anything to do with politics.  But, there are some things that are out of my hands.  

I use Dexcom products because they are the best for my needs, not because I agree with the political views of the heads of Dexcom.

I am not the lead developer.  
The first step is for someone to develop this.  Then, I will ask questions and state the facts that there is already too much in xDrip and I cannot be the only one who spends time fixing bugs.  Those who add features should do some of that work.  That's only fair.

Then, the lead developer will decide to merge or not.
